src/parse/errors.rs

```diff
@@ -440,604 +436,415 @@ impl Error {
-    ) -> io::Result<Self>
-    where
-        V: AsRef<str> + Display + ToOwned,
-        U: Display,
-    {
-        let v = val.as_ref();
+    pub(crate) fn too_many_values(val: &str, arg: &Arg, usage: &str, color: ColorChoice) -> Self {
         let mut c = Colorizer::new(true, color);
 
-        start_error(&mut c, "The value '")?;
-        c.warning(v)?;
-        c.none("' was provided to '")?;
-        c.warning(&arg.to_string())?;
-        c.none("' but it wasn't expecting any more values")?;
-        put_usage(&mut c, usage)?;
-        try_help(&mut c)?;
-
-        Ok(Error {
-            cause: format!(
-                "The value '{}' was provided to '{}', but it wasn't expecting any more values",
-                v, arg
-            ),
+        start_error(&mut c, "The value '");
+        c.warning(val);
+        c.none("' was provided to '");
+        c.warning(arg.name);
+        c.none("' but it wasn't expecting any more values");
+        put_usage(&mut c, usage);
+        try_help(&mut c);
+
+        Error {
             message: c,
             kind: ErrorKind::TooManyValues,
-            info: Some(vec![arg.name.to_owned(), v.to_owned()]),
-        })
+            info: vec![arg.to_string(), val.to_string()],
+        }
     }
 
-    pub(crate) fn too_few_values<U>(
+    pub(crate) fn too_few_values(
         arg: &Arg,
         min_vals: u64,
         curr_vals: usize,
-        usage: U,
+        usage: &str,
         color: ColorChoice,
-    ) -> io::Result<Self>
-    where
-        U: Display,
-    {
+    ) -> Self {
         let mut c = Colorizer::new(true, color);
         let verb = Error::singular_or_plural(curr_vals);
 
-        start_error(&mut c, "The argument '")?;
-        c.warning(&arg.to_string())?;
-        c.none("' requires at least ")?;
-        c.warning(&min_vals.to_string())?;
-        c.none(" values, but only ")?;
-        c.warning(&curr_vals.to_string())?;
-        c.none(&format!(" {} provided", verb))?;
-        put_usage(&mut c, usage)?;
-        try_help(&mut c)?;
-
-        Ok(Error {
-            cause: format!(
-                "The argument '{}' requires at least {} values, but only {} {} provided",
-                arg, min_vals, curr_vals, verb
-            ),
+        start_error(&mut c, "The argument '");
+        c.warning(arg.name);
+        c.none("' requires at least ");
+        c.warning(&min_vals.to_string());
+        c.none(" values, but only ");
+        c.warning(&curr_vals.to_string());
+        c.none(format!(" {} provided", verb));
+        put_usage(&mut c, usage);
+        try_help(&mut c);
+
+        Error {
             message: c,
             kind: ErrorKind::TooFewValues,
-            info: Some(vec![arg.name.to_owned()]),
-        })
+            info: vec![arg.to_string(), curr_vals.to_string(), min_vals.to_string()],
+        }
     }
 
     pub(crate) fn value_validation(
-        arg: Option<&Arg>,
-        err: &str,
+        arg: String,
```

Comment:
Why is this not &Arg? Why is `val` and `err` not `&str`?

Reply:
Because we don't always have the `Arg`, like in `value_of_t`.